spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/DefaultSubscriberFactory.java

```diff
@@ -34,37 +37,130 @@
-		Assert.notNull(executorProvider, "The executor provider can't be null.");
-		Assert.notNull(channelProvider, "The channel provider can't be null.");
-		Assert.notNull(credentialsProvider, "The credentials provider can't be null.");
 
 		this.projectId = projectIdProvider.getProjectId();
 		Assert.hasText(this.projectId, "The project ID can't be null or empty.");
+	}
+
+	/**
+	 * @param executorProvider provides the executor to be used by the subscriber, useful to specify
+	 *                         the number of threads to be used by each executor
+	 */
+	public void setExecutorProvider(ExecutorProvider executorProvider) {
 		this.executorProvider = executorProvider;
+	}
+
+	/**
+	 * @param channelProvider provides the channel to be used by the subscriber
+	 */
+	public void setChannelProvider(TransportChannelProvider channelProvider) {
 		this.channelProvider = channelProvider;
+	}
+
+	/**
+	 * @param credentialsProvider provides the GCP credentials to be used by the subscriber on the
+	 *                            API calls it makes
+	 */
+	public void setCredentialsProvider(CredentialsProvider credentialsProvider) {
 		this.credentialsProvider = credentialsProvider;
 	}
 
+	/**
+	 * @param headerProvider provides headers for the API's HTTP calls
+	 */
+	public void setHeaderProvider(HeaderProvider headerProvider) {
+		this.headerProvider = headerProvider;
+	}
+
+	/**
+	 * @param systemExecutorProvider provides an executor for polling and managing lease extensions
+	 */
+	public void setSystemExecutorProvider(ExecutorProvider systemExecutorProvider) {
+		this.systemExecutorProvider = systemExecutorProvider;
+	}
+
+	/**
+	 * @param flowControlSettings configures the flow control of the subscriber, including the
+	 *                            behaviour for when flow limits are hit
+	 */
+	public void setFlowControlSettings(FlowControlSettings flowControlSettings) {
+		this.flowControlSettings = flowControlSettings;
+	}
+
+	/**
+	 * @param maxAckDurationPeriod sets the maximum period the ack period is extended by
+	 */
+	public void setMaxAckDurationPeriod(Duration maxAckDurationPeriod) {
+		this.maxAckDurationPeriod = maxAckDurationPeriod;
+	}
+
+	/**
+	 * @param parallelPullCount sets the number of pull workers
+	 */
+	public void setParallelPullCount(Integer parallelPullCount) {
+		this.parallelPullCount = parallelPullCount;
+	}
+
 	@Override
 	public Subscriber getSubscriber(String subscriptionName, MessageReceiver receiver) {
-		return Subscriber.newBuilder(
-				SubscriptionName.of(this.projectId, subscriptionName), receiver)
-				.setChannelProvider(this.channelProvider)
-				.setExecutorProvider(this.executorProvider)
-				.setCredentialsProvider(this.credentialsProvider)
-				.build();
+		Subscriber.Builder subscriberBuilder = Subscriber.newBuilder(
+				SubscriptionName.of(this.projectId, subscriptionName), receiver);
+
+		if (this.channelProvider != null) {
+			subscriberBuilder.setChannelProvider(this.channelProvider);
```

Comment:
Again, all of this can be simplified if you do the builder configuration in the setters.